blockchain/utxoviewpoint.go

```diff
@@ -17,184 +17,169 @@ import (
+type utxoFlags uint8
+
+const (
+	// utxoFlagCoinBase indicates that a txout was contained in a coinbase tx.
+	utxoFlagCoinBase utxoFlags = 1 << iota
+
+	// utxoFlagSpent indicates that a txout is spent.
+	utxoFlagSpent
+
+	// utxoFlagModified indicates that a txout has been modified since it was
+	// loaded.
+	utxoFlagModified
+
+	// utxoFlagHasExpiry indicates that a txout was contained in a tx that
+	// included an expiry.
+	utxoFlagHasExpiry
+)
 
-// maybeDecompress decompresses the amount and public key script fields of the
-// utxo and marks it decompressed if needed.
-func (o *utxoOutput) maybeDecompress(compressionVersion uint32) {
-	// Nothing to do if it's not compressed.
-	if !o.compressed {
-		return
+const (
+	// utxoFlagTxTypeBitmask describes the bitmask that yields bits 4-7 from
+	// utxoFlags.
+	utxoFlagTxTypeBitmask = 0xf0
+
+	// utxoFlagTxTypeShift is the number of bits to shift utxoFlags to the right
+	// to yield the correct integer value after applying the bitmask with AND.
+	utxoFlagTxTypeShift = 4
+)
+
+// encodeUtxoFlags returns utxoFlags representing the passed parameters.
+func encodeUtxoFlags(coinbase bool, spent bool, modified bool, hasExpiry bool,
+	txType stake.TxType) utxoFlags {
+
+	packedFlags := utxoFlags(txType) << utxoFlagTxTypeShift
+	if coinbase {
+		packedFlags |= utxoFlagCoinBase
+	}
+	if spent {
+		packedFlags |= utxoFlagSpent
+	}
+	if modified {
+		packedFlags |= utxoFlagModified
+	}
+	if hasExpiry {
+		packedFlags |= utxoFlagHasExpiry
 	}
 
-	o.pkScript = decompressScript(o.pkScript, compressionVersion)
-	o.compressed = false
+	return packedFlags
 }
 
-// UtxoEntry contains contextual information about an unspent transaction such
-// as whether or not it is a coinbase transaction, which block it was found in,
-// and the spent status of its outputs.
+// UtxoEntry houses details about an individual transaction output in a utxo
+// view such as whether or not it was contained in a coinbase tx, the height of
+// the block that contains the tx, whether or not it is spent, its public key
+// script, and how much it pays.
 //
 // The struct is aligned for memory efficiency.
 type UtxoEntry struct {
-	sparseOutputs map[uint32]*utxoOutput // Sparse map of unspent outputs.
-	stakeExtra    []byte                 // Extra data for the staking system.
+	amount   int64
+	pkScript []byte
 
-	txType    stake.TxType // The stake type of the transaction.
-	height    uint32       // Height of block containing tx.
-	index     uint32       // Index of containing tx in block.
-	txVersion uint16       // The tx version of this tx.
+	// ticketMinOuts is the minimal outputs for the ticket transaction that the
+	// output is contained in.  This is only stored in ticket submission outputs
+	// and is nil for all other output types.
+	//
+	// This approach is used since all outputs of a ticket need to be retrieved
+	// when validating vote transaction inputs or revoking a ticket.
+	ticketMinOuts []byte
```

Comment:
We discussed this a bit separately but adding some detail here for reference.

I considered an alternative approach to this where we would store all ticket outputs in the view (as separate utxo entries) and then retrieve them all as needed (when validating vote transaction inputs or revoking a ticket).  This would minimize the duplication of data being stored, but it led to some complexity for a few reasons:

- You need to add all ticket outputs to the view any time you add a ticket submission output to the view (which also requires knowing the number of outputs for the ticket, which would need to be stored for both unspent and spent outputs).
- You need to store ticket commitments as unspent and spent outputs (and even serialize them to spend journal entries as spent), which doesn't really make sense since they aren't truly spendable.

To avoid this taking 24 bytes when its `nil` (which will be the case for all outputs other than ticket submissions), I updated this to be a pointer to a struct that holds the data (and therefore will reduce the usage to 8 bytes for a `nil` pointer).  I went with a struct here rather than a pointer to a slice directly since we will likely want to expand what is stored along with that data when implementing the utxo cache.
```
func main() {
	type ticketMinimalOutputs struct {
		data []byte
	}
	type UtxoEntry struct {
		ticketMinOuts *ticketMinimalOutputs
	}
	entry := UtxoEntry{ticketMinOuts: nil}
	fmt.Println(unsafe.Sizeof(entry))
}
```
```
$ go run main.go
8
```

This still leaves plenty of room for improvement, and I'll consider some alternative approaches as part of the utxo cache work.  I think ultimately, we will want separate views for the regular and stake trees, which we have discussed previously and should allow us to simplify some scenarios like that.